### ISSUE TITLE
Add support for ARRAY concatenation

### ISSUE THREAD

**Issue by lukaseder** [jOOQ/jOOQ#14388]
Similar to `JSON` and `JSONB` concatenation (see https://github.com/jOOQ/jOOQ/issues/13950), we should support concatenating arrays, e.g. using

- [x] PostgreSQL: `||`
- [x] H2: `||`
- [x] HSQLDB: `||`

A very laborious emulation seems possible in other dialects in the future for `a || b`:

```sql
select case when a is not null and b is not null then coalesce((
  select array_agg(x order by a, o)
  from (
    select 1 a, x, o 
    from unnest(a) with ordinality as t(x, o) 
    union all 
    select 2 a, x, o 
    from unnest(b) with ordinality as t(x, o)
  ), a)
) end
from (values (array[1, 2], array[3, 4])) as t (a, b);
```

Producing the expected:

```
|C1       |
|---------|
|[1,2,3,4]|
```

It should be possible to concatenate any pairs of:

- `T[]` and `T[]` types

These ones look useful, but can't be reasonably modeled in the type system, given erasure:

- `T` and `T[]` types
- `T[]` and `T` types

This will be implemented as a new `DSL::arrayConcat` / `QOM.ArrayConcat` function. While it may be tempting to retrofit the `Field::concat` function, that one isn't generic yet, so it would be a breaking change.

**Comment by lukaseder** [comment]
Actually, HSQLDB also supports `||`. I ran into an unrelated syntax error and wrongly assumed it wouldn't support the syntax. 😅

**Comment by lukaseder** [comment]
Implemented for jOOQ 3.18.0
